com.unity.render-pipelines.universal/ShaderLibrary/Lighting.hlsl

```diff
@@ -271,14 +333,35 @@ half OneMinusReflectivityMetallic(half metallic)
 {
     // We'll need oneMinusReflectivity, so
     //   1-reflectivity = 1-lerp(dielectricSpec, 1, metallic) = lerp(1-dielectricSpec, 0, metallic)
-    // store (1-dielectricSpec) in kDieletricSpec.a, then
+    // store (1-dielectricSpec) in kDielectricSpec.a, then
     //   1-reflectivity = lerp(alpha, 0, metallic) = alpha + metallic*(0 - alpha) =
     //                  = alpha - metallic * alpha
-    half oneMinusDielectricSpec = kDieletricSpec.a;
+    half oneMinusDielectricSpec = kDielectricSpec.a;
     return oneMinusDielectricSpec - metallic * oneMinusDielectricSpec;
 }
 
-inline void InitializeBRDFData(half3 albedo, half metallic, half3 specular, half smoothness, half alpha, out BRDFData outBRDFData)
+#ifdef _CLEARCOAT
+
+// See ConvertF0ForAirInterfaceToF0ForClearCoat15
+half3 f0ClearCoatToSurface(half3 f0)
+{
+    // Approximation of iorTof0(f0ToIor(f0), 1.5)
+    // This assumes that the clear coat layer has an IOR of 1.5
+#if defined(SHADER_API_MOBILE)
```

Comment:
be careful as you will conflcit with the core Bsdf function
why not use it ?

TEMPLATE_1_REAL(ConvertF0ForAirInterfaceToF0ForClearCoat15, fresnel0, return saturate(-0.0256868 + fresnel0 * (0.326846 + (0.978946 - 0.283835 * fresnel0) * fresnel0)))

you could simply define the mobile version there too

#if defined(SHADER_API_MOBILE)
TEMPLATE_1_REAL(ConvertF0ForAirInterfaceToF0ForClearCoat15, fresnel0, rf0 * (f0 * 0.526868h + 0.529324h) - 0.0482256h)
#else
TEMPLATE_1_REAL(ConvertF0ForAirInterfaceToF0ForClearCoat15, fresnel0, return saturate(-0.0256868 + fresnel0 * (0.326846 + (0.978946 - 0.283835 * fresnel0) * fresnel0)))
#endif

Reply:
Yes, this is basically the same function (I plotted it out just to be sure). It's origins are from the integrated code from UTJ / LWRP. Basically it's a half version of ConvertF0ForAirInterfaceToF0ForClearCoat15 with the SHADER_API_MOBILE approximation.

I discussed this with @phi-lira  and we thought at least to rename it to ConvertF0ForAirInterfaceToF0ForClearCoat15Half.
 
I definitely would like to merge this with the Core function if that is ok? 

I would like to keep the "halfness" of the function though. Looking at the core/ShaderLibrary/Common.hlsl it looks like it will be half on mobile.  
`#if defined(SHADER_API_MOBILE) || defined(SHADER_API_SWITCH)
#define HAS_HALF 1
#else
#define HAS_HALF 0
#endif
`
since half is now available on new desktop hw too I wonder if we should try to support/prepare for that too?

Reply:
MY own thought about the usage of half on desktop is that it will be use in particular code and optimize for this case. On desktop it don't really make sense to have a random function using half rather than float in particular for a BRDF, but it is way more efficient to have a loop with hundred of ALU doig some heavy calculation using half. So my guess is that half will be restricted usage and we don't need to care about it for what we are doing. So not sure we need a dedicated ConvertF0ForAirInterfaceToF0ForClearCoat15Half as it will just end in the code as

Do clear coat:

Foo()
{...
#if defined(SHADER_API_MOBILE) || defined(SHADER_API_SWITCH)
ConvertF0ForAirInterfaceToF0ForClearCoat15
#else
ConvertF0ForAirInterfaceToF0ForClearCoat15Half
#endif
}

So we will pollute the code for no good reasons.

However it is just an assumption, I am open to counter argument, or if you think there is a good reasons to have both half and float version, could be ok. But in this case we shouldn't do half vs float as it is not the goal of this function here, but rather a fast vs regular approx:

FastConvertF0ForAirInterfaceToF0ForClearCoat15

and 

ConvertF0ForAirInterfaceToF0ForClearCoat15

This will make more sens to me.

Thought?

Reply:
Yes, I agree, and would expect, that most code for desktop half would be a complete feature/shader (when you don't need the precision). I haven't really been able to use desktop half in practice yet, so I don't know if there will be wins for having couple lines here and there as half in ALU bound shader. Or is it like SIMD on CPU that you need a good chunk of code (lots of instructions) for it be a real win. Then again tactical use of explicit half functions (where precision allows) can add up into a longer pieces of half code even if the "feature"/shader isn't written half in mind (just like SIMD on CPU).
I would either move the approx under SHADER_API_MOBILE define into the BSDF.hlsl ConvertF0ForAirInterfaceToF0ForClearCoat15() or add the approx as a separate function (maybe only URP side) for mobile and just use it in mobile half code. It's only 2-3 mul/add less so probably doesn't matter too much on desktop.

Reply:
I changed this to use the in BSDF.hlsl for desktop and I added a separate function for the approximation which is used for mobile. It's approx is URP only code for now, but it's separated so it can be easily moved later. The define part is meant to be left in URP side.
I did not want to move the approx inside a define into BSDF.hlsl as I didn't want to impact quality behind everyone's back.